### PR TITLE
💥 Register transactions as blobs

### PR DESCRIPTION
**Pull request by wraitii**

### CODE REVIEW

File: crates/client-sdk/src/rest_client.rs [hunk 1]
```diff
@@ -32,13 +32,15 @@ impl NodeApiHttpClient {
...
 
+    /*
     pub async fn send_tx_register_contract(
         &self,
         tx: &RegisterContractTransaction,
     ) -> Result<TxHash> {
         self.post("v1/contract/register", tx, "Sending tx register contract")
             .await
     }
+    */
```
Comment (wraitii) [inline crates/client-sdk/src/rest_client.rs:43]: Est-ce que je le garde?

File: src/node_state.rs [hunk 2]
```diff
@@ -208,18 +214,18 @@ impl NodeState {
...
             .collect();
 
-        self.unsettled_transactions.add(UnsettledBlobTransaction {
+        should_try_and_settle = self.unsettled_transactions.add(UnsettledBlobTransaction {
             identity: tx.identity.clone(),
             hash: blob_tx_hash.clone(),
             blobs_hash,
             blobs,
-        });
+        }) && should_try_and_settle;
```
Comment (BertrandD) [inline src/node_state.rs:222]: J'ai repris mon commentaire 3 fois car c'est très implicite ce qui se passe ici...

Je pense que ce serait plus clair de changer quelques lignes plus haut:

```rust
                } else if blob.contract_name.0 != "hyle" {
                    should_try_and_settle = false;
                }
```
en
```rust
                } else {
                    should_try_and_settle = should_try_and_settle && (blob.contract_name.0 == "hyle");
                }
```

et ne pas changer la fonction `add`.

Comment (BertrandD) [reply src/node_state.rs:222]: ah non en effet, si on a une BlobTx qui passe avec un registerContract et des zkBlobs , puis une 2e BlobTx avec uniquement un register contract, il faut attendre que la Tx précédente soit settle...

Ce qui veut dire que tu peux bloquer tous les register contract "futurs" si tu settle jamais ta Tx, les autres ne pourront être settle qu'à la fin du timeout... On peut facilement DOS le truc en envoyant des Tx de register avec un autre blob à côté qu'on prouve jamais, ce qui rend les register contract "lent"...

IMHO on devrait pas pouvoir faire autre chose que register des contract dans une tx de register contract, mais en même temps, je dis ça et je me dis qu'on voudra sûrement composer le register contract avec une vérification d'identité.... Ce truc de timeout c'est relou...

TLDR: tout va bien, let's go comme ça

File: crates/hyle-model/src/contract.rs [hunk 3]
```diff
@@ -315,3 +316,41 @@ impl Add<BlockHeight> for BlockHeight {
...
+impl ContractAction for RegisterContractAction {
+    fn as_blob(
+        &self,
+        contract_name: ContractName,
+        caller: Option<BlobIndex>,
+        callees: Option<Vec<BlobIndex>>,
+    ) -> Blob {
+        Blob {
+            contract_name,
+            data: BlobData::from(StructuredBlobData {
```
Comment (BertrandD) [inline crates/hyle-model/src/contract.rs:349]: On veut vraiment utiliser un StructuredBlobData ici ? Je vois pas l'usage

File: src/node_state.rs [hunk 4]
```diff
@@ -508,6 +548,46 @@ impl NodeState {
...
+        let Ok(reg) =
+            StructuredBlobData::<RegisterContractAction>::try_from(current_blob.data.clone())
+        else {
+            bail!("Blob is  not a RegisterContractAction");
+        };
+        // Check name, it's either a direct subdomain or a TLD
+        let parts = reg.parameters.contract_name.0.split(".");
+        let is_subdomain =
+            parts.clone().count() == 2 && parts.last().map(|s| s == "hyle").unwrap_or(false);
+        let is_tld = !reg.parameters.contract_name.0.contains(".");
```
Comment (maxgttph) [inline src/node_state.rs:565]: Pas sûr que ça soit nécessaire, puisqu'on vérifie déjà `parts.clone().count() == 2`

File: crates/hyle-model/src/contract.rs (hunk 3 above)
Comment (BertrandD) [reply crates/hyle-model/src/contract.rs:349]: IMHO on devrait avoir 2 traits: 

```rust
trait StructuredBlobContractAction {
    fn as_blob(
        &self,
        contract_name: ContractName,
        caller: Option<BlobIndex>,
        callees: Option<Vec<BlobIndex>>,
    ) -> Blob;
}

trait SimpleBlobContractAction {
    fn as_blob(&self, contract_name: ContractName) -> Blob;
}
```

mais je sais pas où ce trait "ContractAction" est requis, on peut peut-être pas faire ça en fait

File: src/node_state.rs (hunk 4 above)
Comment (maxgttph) [reply src/node_state.rs:565]: D'ailleurs est-ce que si on register ".hyle" on a le bon comportement ?

File: src/node_state.rs [hunk 5]
```diff
@@ -121,28 +144,13 @@ impl NodeState {
...
 
-    pub fn handle_register_contract_tx(
-        &mut self,
-        tx: &RegisterContractTransaction,
-    ) -> Result<(), Error> {
-        if self.contracts.contains_key(&tx.contract_name) {
-            bail!("Contract already exists")
-        }
+    pub fn handle_register_contract_action(&mut self, tx: &RegisterContractAction) {
         info!("📝 Registering new contract {}", tx.contract_name);
```
[the remark this answers is not in the document]
Comment (wraitii) [reply src/node_state.rs:154]: Si on l'update ça sera ici aussi, mais je peux changer le message

File: src/utils/integration_test.rs [hunk 6]
```diff
@@ -254,27 +257,60 @@ impl NodeIntegrationCtx {
 
         let mut handler = ModulesHandler::new(&bus).await;
 
-        Self::build_module::<Mempool>(&mut handler, &ctx, &ctx, &mut mocks).await?;
+        Self::build_module::<Mempool>(&mut handler, &ctx, ctx.clone(), &mut mocks).await?;
 
-        Self::build_module::<Genesis>(&mut handler, &ctx, &ctx, &mut mocks).await?;
+        Self::build_module::<Genesis>(&mut handler, &ctx, ctx.clone(), &mut mocks).await?;
```
Comment (BertrandD) [inline src/utils/integration_test.rs:262]: pourquoi faire les clone là plutôt que dans la fonction ? en vrai osef on est dans les tests, mais c'est moins verbeux de faire le clone dans la fonction non ?

File: src/node_state.rs (hunk 4 above)
Comment (wraitii) [reply src/node_state.rs:565]: C'pas le même check, mais bonne idée de test

File: tests/amm_test.rs [hunk 7]
```diff
@@ -132,7 +132,7 @@ mod e2e_amm {
         ///////////////////// hyllar2 contract registration /////////////////
         info!("➡️  Registring hyllar2 contract");
         const HYLLAR2_CONTRACT_NAME: &str = "hyllar2";
-        ctx.register_contract::<HyllarContract>(HYLLAR2_CONTRACT_NAME)
+        ctx.register_contract::<HyllarContract>("hyle.hyle".into(), HYLLAR2_CONTRACT_NAME)
```
Comment (BertrandD) [inline tests/amm_test.rs:135]: si ça a vocation à toujours être la même valeur, on peut pas le mettre en dur dans `register_contract` ?

File: src/mempool.rs [hunk 8]
```diff
@@ -62,15 +62,12 @@ impl KnownContracts {
         contract_name: &ContractName,
         verifier: &Verifier,
         program_id: &ProgramId,
-    ) -> Result<()> {
-        if self.0.contains_key(contract_name) {
-            bail!("Contract {contract_name} already exists")
-        }
+    ) {
+        debug!("🏊📝 Registering contract in mempool {:?}", contract_name);
```
[the remark this answers is not in the document]
Comment (wraitii) [reply src/mempool.rs:66]: J'ai juste supprimé le `new`, mais idem c'est potentiellement une update

File: src/mempool.rs [hunk 9]
```diff
@@ -1425,16 +1422,17 @@ pub mod test {
     }
 
     pub fn make_register_contract_tx(name: ContractName) -> Transaction {
-        Transaction {
-            version: 1,
-            transaction_data: TransactionData::RegisterContract(RegisterContractTransaction {
-                owner: "test".to_string(),
+        BlobTransaction {
+            identity: "hyle.hyle".into(),
+            blobs: vec![RegisterContractAction {
```
Comment (BertrandD) [inline src/mempool.rs:1427]: je propose une fonction 

```rust
sdk::helpers::register_contract(ContractName, ProgramId, StateDigest, Verifier) -> BlobTransaction
```

File: src/consensus.rs [hunk 10]
```diff
@@ -698,13 +698,15 @@ impl Consensus {
                 for action in block.staking_actions {
                     match action {
                         (identity, StakingAction::Stake { amount }) => {
+                            debug!("🔒 Staking {} for {}", amount, identity);
```
[the remark this answers is not in the document]
Comment (wraitii) [reply src/consensus.rs:701]: Oui je les ai remis en debug pour différencier de ceux que j'ai dans le proving mais ils sont pas stricto sensu nécessaire, je peux les resupprimer

File: src/node_state.rs (hunk 2 above)
Comment (wraitii) [reply src/node_state.rs:222]: Oui c'est le problème et l'avantage - fait avec des blobs j'ai toute la comptabilité à ma disposition pour faire des logiques de registration fancy, mais on bloque sur les timeouts...
En pratique si on early-settle les "success = failed" ça me semble fine

File: crates/hyle-model/src/contract.rs (hunk 3 above)
Comment (wraitii) [reply crates/hyle-model/src/contract.rs:349]: J'ai utilisé structuredBlob parce que comme ça si quelqu'un veut faire une fonction qui appelle "register_contract" dans un autre contrat c'est compatible et ça me semble worth it (et staking fait ça je crois).

Idéalement on pourrait être compatible avec ou sans caller, mais la façon dont on a fait les structuredblob le permet pas en l'état.

File: src/utils/integration_test.rs (hunk 6 above)
Comment (wraitii) [reply src/utils/integration_test.rs:262]: Ça m'oblige à mettre Clone comme trait et RestApiContext est pas Clone (et c'est une tannée de le transformer - ou alors je l'Arc mais pour rien)

File: tests/amm_test.rs (hunk 7 above)
Comment (wraitii) [reply tests/amm_test.rs:135]: Ça a pas vocation à être toujours la même valeur justement, c'est juste que dans les tests basiques j'ai pas mis de vérification d'identité mais j'ai un test qui le fait plus bas.

File: src/node_state.rs [hunk 11]
```diff
@@ -38,6 +38,29 @@ pub struct NodeState {
     unsettled_transactions: OrderedTxMap,
 }
 
+// TODO: change this.
```
[the remark this answers is not in the document]
Comment (wraitii) [reply src/node_state.rs:41]: Mh non c'est juste extrêmement peu clair. L'idée c'est que `hyle` devrait être register dans le genesis block et pas hardocdé

File: crates/hyle-model/src/contract.rs (hunk 3 above)
Comment (BertrandD) [reply crates/hyle-model/src/contract.rs:349]: > une fonction qui appelle "register_contract" dans un autre contrat c'est compatible

je comprends pas :thinking: 

dans staking j'ai utilisé un Structuredblob en prévision des opérations de claim rewards si je me souviens bien

File: src/utils/integration_test.rs (hunk 6 above)
Comment (BertrandD) [reply src/utils/integration_test.rs:262]: oh ok !

File: crates/hyle-model/src/contract.rs (hunk 3 above)
Comment (wraitii) [reply crates/hyle-model/src/contract.rs:349]: Je veux dire tu peux envisager un contrat qui appelle un contrat pour register un nouveau domain-name, et l'identité c'est le caller contract, pas `identity`. Avoir les caller/callees ça me semble worth it

File: src/mempool.rs (hunk 9 above)
Comment (wraitii) [reply src/mempool.rs:1420]: Vais plutôt le faire dans le transaction_builder et dans une autre PR je pense

Comment (BertrandD) [reply src/mempool.rs:1420]: pour quand tu le fera: je pense que ce serait bien si ça peut être aussi accessible sans le tx builder. Il y a des exemples où je n'utilise pas le tx builder

File: src/model/contract_registration.rs [hunk 12]
```diff
@@ -0,0 +1,152 @@
+use anyhow::{bail, Result};
+use hyle_model::{BlobTransaction, ContractName, RegisterContractAction, StructuredBlobData};
+
+/// Check that the new contract name is:
+/// - a valid subdomain of the owner contract name.
+/// - the exact same domain (for updating the contract).
+pub fn validate_contract_registration(
```
Comment (BertrandD) [inline src/model/contract_registration.rs:7]: j'ai été étonné de voir ça dans le model, mais je comprends qu'on puisse en avoir besoin en dehors de la node. Ca ça être la même chose pour toute les fonction de "validation" de la donnée...

Est-ce que ce serait pas mieux en impl de ContractName ?

```rust
impl ContractName {
  pub fn validate(&self, owner: ContractName) {...}
}
```

File: src/model/contract_registration.rs [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+            "Invalid subdomain contract name for '{}': {}",
+            owner,
+            new_contract_name
+        );
+    }
+    Ok(())
+}
+
+// TODO: would be nice to not clone
+pub fn validate_any_contract_registration(tx: &BlobTransaction) -> Result<()> {
```
Comment (BertrandD) [inline src/model/contract_registration.rs:48]: same ? impl BlobTransaction comme le validate identity ?

https://github.com/Hyle-org/hyle/blob/53226174b2a40647e4735f44657e7dca2eaa1045/crates/hyle-model/src/transaction.rs#L198-L204

File: src/model/contract_registration.rs (hunk 12 above)
Comment (BertrandD) [reply src/model/contract_registration.rs:7]: ah lol on est pas dans la crate model :D sorry XD j'ai lu trop vite

File: src/model/contract_registration.rs [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+        {
+            return Ok(());
+        }
+    } else if owner == new_contract_name && !owner.0.is_empty() {
+        return Ok(());
+    }
+
+    let Some((name, tld)) = new_contract_name.0.split_once(".") else {
+        bail!(
+            "Invalid contract name for '{}': {}",
```
Comment (BertrandD) [inline src/model/contract_registration.rs:25]: ce serait bien d'avoir des messages d'erreur différent à chaque fois qui dise quelle est la condition invalide

File: src/model/contract_registration.rs [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+                &blob.contract_name,
+                &reg.parameters.contract_name,
+            );
+        }
+        Ok(())
+    })
+}
+
+#[cfg(test)]
+mod test {
```
Comment (BertrandD) [inline src/model/contract_registration.rs:61]: typiquement là on devrait utiliser du proptest pour bien tester les edge cases

**Review by BertrandD — APPROVED** [review]

File: src/model/contract_registration.rs (hunk 12 above)
Comment (wraitii) [reply src/model/contract_registration.rs:7]: Ouais je savais pas trop où le mettre, c'est pas vraiment un `utils`. Je l'ai mis ici un peu au pif.

Comment (wraitii) [reply src/model/contract_registration.rs:7]: Ça me semble être une logique un peu trop spécifique à la node pour le mettre en `impl` direct mais ça se débat

File: src/model/contract_registration.rs (hunk 15 above)
Comment (wraitii) [reply src/model/contract_registration.rs:61]: ouais j'ai eu un peu la flemme de refaire ça mais ça serait intéressant.

File: src/mempool.rs (hunk 9 above)
Comment (wraitii) [reply src/mempool.rs:1420]: En vrai c'est pas _si_ compliqué de créer une blobTransaction et l'action non plus, je trouve 😇

File: src/model/contract_registration.rs (hunk 14 above)
Comment (wraitii) [reply src/model/contract_registration.rs]: Done
